src/Illuminate/Console/Signals.php

```diff
@@ -0,0 +1,132 @@
+{
+    /**
+     * The signal registry instance.
+     *
+     * @var \Symfony\Component\Console\SignalRegistry\SignalRegistry
+     */
+    protected $registry;
+
+    /**
+     * The signal registry's previous list of handlers.
+     *
+     * @param array<int, array<int, callable>>|null
+     */
+    protected $previousHandlers;
+
+    /**
+     * Create a new Signals instance.
+     *
+     * @param  \Symfony\Component\Console\SignalRegistry\SignalRegistry  $registry
+     * @return void
+     */
+    public function __construct($registry)
+    {
+        $this->registry = $registry;
+
+        $this->previousHandlers = $this->getHandlers();
+    }
+
+    /**
+     * Registers a new signal handler.
+     *
+     * @param  int  $signal
+     * @param  callable  $callback
+     * @return void
+     */
+    public function register($signal, $callback)
+    {
+        $this->previousHandlers[$signal] ??= $this->initializeSignal($signal);
+
+        with($this->getHandlers(), function ($handlers) use ($signal) {
+            $handlers[$signal] ??= $this->initializeSignal($signal);
+
+            $this->setHandlers($handlers);
+        });
+
+        $this->registry->register($signal, $callback);
+
+        with($this->getHandlers(), function ($handlers) use ($signal) {
+            $lastHandlerInserted = array_pop($handlers[$signal]);
+            array_unshift($handlers[$signal], $lastHandlerInserted);
+
+            $this->setHandlers($handlers);
+        });
+    }
+
+    /**
+     * Unregister the current signals instance, and reverts
+     * the signal's registry handlers state.
+     *
+     * @return array<int, array<int, callable>>
+     */
+    public function unregister()
+    {
+        $this->setHandlers($this->previousHandlers);
+    }
+
+    /**
+     * Executes the given callback if "signals" should be used and are available.
+     *
+     * @param  callable  $callback
+     * @return void
+     */
+    public static function whenAvailable($callback)
+    {
+        if (
+            app()->runningInConsole()
+            && ! app()->runningUnitTests()
+            && extension_loaded('pcntl')
+            && defined('SIGINT')
+            && SignalRegistry::isSupported()) {
+            $callback();
+        }
+    }
+
+    /**
+     * Set the registry's handlers.
+     *
+     * @param  array<int, array<int, callable>>  $handlers
+     * @return void
+     */
+    protected function setHandlers($handlers)
+    {
+        (fn () => $this->signalHandlers = $handlers)
+            ->call($this->registry);
+    }
+
+    /**
+     * Get the registry's handlers.
+     *
+     * @return array<int, array<int, callable>>
+     */
+    protected function getHandlers()
+    {
+        return (fn () => $this->signalHandlers)
+            ->call($this->registry);
+    }
+
+    /**
+     * Gets the signal's default callback on the array format.
+     *
+     * @return [callable]
+     */
+    protected function initializeSignal($signal)
```

Comment:
@taylorotwell @driesvints @GrahamCampbell Undecided about this peace of code, how it should behave in different signals. For the context, by default, PHP ends the script when receiving a SIGTERM or similar. Yet, when you override the signal's handler, PHP actually does not end the script unless you call an `exit()`. 

Yet, as a user, my expectation when using something like `$this->trap(SIGTERM, fn () => @unlink($file));` is that PHP ends the script after my handler is done. Therefore, the code bellow helps here.

What are your thoughts?

Reply:
Like you, my expectation is that I want the script to end.